geomstats/sphere.py

```diff
@@ -0,0 +1,114 @@
+"""
+Computations on the n-dimensional sphere
+embedded in the (n+1)-dimensional Euclidean space.
+"""
+
+import numpy as np
+
+EPSILON = 1e-6
+
+
+def projection_to_tangent_space(ref_point, vector):
+    """
+    Project the vector v onto the tangent space:
+    T_{ref_point} S = {w | scal(w, ref_point) = 0}
+    """
+    sq_norm = np.dot(ref_point, ref_point)
+    tangent_vec = vector - np.dot(ref_point, vector) / sq_norm * ref_point
+    return tangent_vec
+
+
+def riemannian_exp(ref_point, vector, epsilon=EPSILON):
+    """
+    Compute the Riemannian exponential at point ref_point
+    of tangent vector tangent_vec wrt the metric obtained by
+    embedding of the n-dimensional sphere
+    in the (n+1)-dimensional euclidean space.
+
+    This gives a point on the n-dimensional sphere.
+
+    :param ref_point: a point on the n-dimensional sphere
+    :param vector: (n+1)-dimensional vector
+    :returns riem_exp: a point on the n-dimensional sphere
+    """
+    tangent_vec = projection_to_tangent_space(ref_point, vector)
+    norm_tangent_vec = np.linalg.norm(tangent_vec)
+
+    if norm_tangent_vec < epsilon:
+        coef_1 = (1. - norm_tangent_vec ** 2 / 2.
+                  + norm_tangent_vec ** 4 / 24.
+                  - norm_tangent_vec ** 6 / 720.
+                  + norm_tangent_vec ** 8 / 40320.)
+        coef_2 = (1. - norm_tangent_vec ** 2 / 6.
```

Comment:
why do you divide by norm_tangent_vec at line 48 and not here?

Reply:
This is because tangent_vector is not normalized, whereas ref_point is (a point on a sphere represented by) a normalized vector.

We could normalize tangent_vector and avoid this, but cos(angle) and sin(angle)/angle show the same orders in the taylor expansions so I'd rather keep them like this.